ibis/pyspark/compiler.py

```diff
@@ -754,3 +822,139 @@ def compile_join(t, expr, scope, how):
     predicates = t.translate(op.predicates[0], scope)
 
     return left_df.join(right_df, predicates, how)
+
+
+@compiles(ops.WindowOp)
+def compile_window_op(t, expr, scope, **kwargs):
+    op = expr.op()
+    window = op.window
+    operand = op.expr
+
+    group_by = window._group_by
+    grouping_keys = [
+        key_op.name
+        if isinstance(key_op, ops.TableColumn)
+        else compile_with_scope(
+            t, key, scope
+        )
+        for key, key_op in zip(
+            group_by, map(operator.methodcaller('op'), group_by)
+        )
+    ]
+
+    order_by = window._order_by
+    ordering_keys = [
+        key.to_expr().get_name()
+        for key in map(operator.methodcaller('op'), order_by)
+    ]
+
+    pyspark_window = Window.partitionBy(grouping_keys).orderBy(ordering_keys)
+    result = t.translate(operand, scope, window=pyspark_window)
+
+    return result
+
+
+@compiles(ops.Lag)
+def compile_lag(t, expr, scope, *, window, **kwargs):
+    op = expr.op()
+
+    src_column = t.translate(op.arg, scope)
+    default = op.default.op().value if op.default is not None else op.default
+    offset = op.offset.op().value if op.offset is not None else op.offset
+
+    if offset:
+        return F.lag(src_column, count=offset, default=default).over(window)
+    else:
+        return F.lag(src_column, default=default).over(window)
+
+
+@compiles(ops.Lead)
+def compile_lead(t, expr, scope, *, window, **kwargs):
+    op = expr.op()
+
+    src_column = t.translate(op.arg, scope)
+    default = op.default.op().value if op.default is not None else op.default
+    offset = op.offset.op().value if op.offset is not None else op.offset
+
+    if offset:
+        return F.lead(src_column, count=offset, default=default).over(window)
+    else:
+        return F.lead(src_column, default=default).over(window)
+
+
+@compiles(ops.MinRank)
+def compile_rank(t, expr, scope, *, window, **kwargs):
+    return F.rank().over(window).astype('long') - F.lit(1)
```

Comment:
is the literal because pyspark starts at `0` and ibis's other backends start at `1`?

Reply:
if so, i'd eventually like to figure out a better contract for these kinds of operations.

Reply:
Yes, the expected result for ibis expressions like `rank()` start indexing at 1 whereas Spark sql starts at 0. I think Pyspark should match the expected ibis behavior (e.g. I wouldn't want a user to get different results from an ibis expression depending on which backend they choose to execute the expression on). If Spark sql `rank()` function changes in the future to index from 1, the unit tests should catch it.

Reply:
It seems index 0 is the default value of a lot of operations (as `rank`) and this `- 1` approach is the current default approach.